File: android/app/src/main/java/me/treexhd/supertrace/traceroute/ReverseDnsModule.kt
package me.treexhd.supertrace.traceroute

import com.facebook.react.bridge.*
import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.atomic.AtomicInteger

class ReverseDnsModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val NAME = "ReverseDnsModule"
        private const val TAG = "SuperTraceRDNS"

        // Public recursive resolvers, ordered by reliability/speed.
        // First match wins; we walk the list per-IP until one answers.
        private val DNS_SERVERS = listOf(
            "1.1.1.1",       // Cloudflare
            "8.8.8.8",       // Google
            "9.9.9.9",       // Quad9
            "1.0.0.1",       // Cloudflare secondary
            "8.8.4.4",       // Google secondary
            "208.67.222.222" // OpenDNS
        )

        private const val DNS_PORT = 53
        // 1500ms per server attempt — tight enough to fall through quickly,
        // generous enough for high-latency mobile links.
        private const val PER_SERVER_TIMEOUT_MS = 1500
        // Hard ceiling per IP across all servers + retries.
        private const val TOTAL_TIMEOUT_MS = 6000
        // Limit concurrent in-flight raw UDP queries so we don't flood the radio.
        private const val MAX_CONCURRENT = 12
    }

    override fun getName(): String = NAME

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val txIdCounter = AtomicInteger((Math.random() * 65535).toInt())
    private val gate = Semaphore(MAX_CONCURRENT)

    @ReactMethod
    fun reverseLookup(ip: String, promise: Promise) {
        scope.launch {
            try {
                val fqdn = withTimeoutOrNull(TOTAL_TIMEOUT_MS.toLong()) {
                    gate.withPermit { performLookup(ip) }
                }
                promise.resolve(fqdn)
            } catch (e: Exception) {
                android.util.Log.w(TAG, "reverseLookup($ip) failed: ${e.message}")
                promise.resolve(null)
            }
        }
    }

    /**
     * Batch lookup — accepts an array of IPs and returns a map { ip: fqdn|null }.
     * Internally throttled by the same semaphore so it can't drown the network.
     */
    @ReactMethod
    fun reverseLookupBatch(ips: ReadableArray, promise: Promise) {
        scope.launch {
            try {
                val ipList = (0 until ips.size()).mapNotNull { ips.getString(it) }
                val results = WritableNativeMap()
                val mutex = kotlinx.coroutines.sync.Mutex()

                coroutineScope {
                    ipList.map { ip ->
                        async {
                            val fqdn = withTimeoutOrNull(TOTAL_TIMEOUT_MS.toLong()) {
                                gate.withPermit { performLookup(ip) }
                            }
                            mutex.lock()
                            try {
                                if (fqdn != null) results.putString(ip, fqdn)
                                else results.putNull(ip)
                            } finally {
                                mutex.unlock()
                            }
                        }
                    }.awaitAll()
                }
                promise.resolve(results)
            } catch (e: Exception) {
                android.util.Log.w(TAG, "reverseLookupBatch failed: ${e.message}")
                promise.resolve(WritableNativeMap())
            }
        }
    }

    /**
     * Try every DNS server in order. First non-empty PTR wins.
     * Each server gets one shot; a NOERROR-with-empty-answer is treated as
     * "no PTR record" and short-circuits — we don't keep asking other servers
     * for an answer that doesn't exist.
     */
    private suspend fun performLookup(ip: String): String? {
        val arpaName = buildArpaName(ip) ?: return null

        for (server in DNS_SERVERS) {
            val outcome = queryPtr(server, arpaName)
            when (outcome) {
                is PtrOutcome.Found -> return outcome.name
                is PtrOutcome.NoRecord -> return null  // authoritative "doesn't exist"
                is PtrOutcome.Failed -> continue       // try next server
            }
        }
        return null
    }

    private sealed class PtrOutcome {
        data class Found(val name: String) : PtrOutcome()
        object NoRecord : PtrOutcome()
        object Failed : PtrOutcome()
    }

    private suspend fun queryPtr(server: String, arpaName: String): PtrOutcome =
        withContext(Dispatchers.IO) {
            var socket: DatagramSocket? = null
            try {
                socket = DatagramSocket()
                socket.soTimeout = PER_SERVER_TIMEOUT_MS

                val txId = (txIdCounter.incrementAndGet() and 0xFFFF)
                val query = buildPtrQuery(txId, arpaName)
                val serverAddr = InetAddress.getByName(server)
                val sendPacket = DatagramPacket(query, query.size, serverAddr, DNS_PORT)
                socket.send(sendPacket)

                val buf = ByteArray(512)
                val recvPacket = DatagramPacket(buf, buf.size)
                socket.receive(recvPacket)

                parsePtrResponse(buf, recvPacket.length, txId)
            } catch (e: Exception) {
                PtrOutcome.Failed
            } finally {
                try { socket?.close() } catch (_: Exception) {}
            }
        }

    /**
     * Translate "1.2.3.4" → "4.3.2.1.in-addr.arpa"
     * or IPv6 "2001:db8::1" → 32 reversed nibbles + ".ip6.arpa"
     */
    private fun buildArpaName(ip: String): String? {
        return try {
            val addr = InetAddress.getByName(ip)
            val bytes = addr.address
            when (bytes.size) {
                4 -> bytes.reversed().joinToString(".") { (it.toInt() and 0xFF).toString() } + ".in-addr.arpa"
                16 -> {
                    val sb = StringBuilder()
                    for (i in 15 downTo 0) {
                        val b = bytes[i].toInt() and 0xFF
                        sb.append(Integer.toHexString(b and 0x0F)).append('.')
                        sb.append(Integer.toHexString((b shr 4) and 0x0F)).append('.')
                    }
                    sb.append("ip6.arpa")
                    sb.toString()
                }
                else -> null
            }
        } catch (e: Exception) {
            null
        }
    }

    private fun buildPtrQuery(txId: Int, arpaName: String): ByteArray {
        val labels = arpaName.split('.').filter { it.isNotEmpty() }
        // 12 byte header + sum(label_len + label) + 1 (root) + 4 (type+class)
        val size = 12 + labels.sumOf { it.length + 1 } + 1 + 4
        val buf = ByteBuffer.allocate(size).order(ByteOrder.BIG_ENDIAN)

        // Header: txId, flags=0x0100 (standard query, RD set), QD=1, AN=0, NS=0, AR=0
        buf.putShort(txId.toShort())
        buf.putShort(0x0100.toShort())
        buf.putShort(1)
        buf.putShort(0)
        buf.putShort(0)
        buf.putShort(0)

        // Question: QNAME
        for (label in labels) {
            val bytes = label.toByteArray(Charsets.US_ASCII)
            buf.put(bytes.size.toByte())
            buf.put(bytes)
        }
        buf.put(0.toByte()) // root

        // QTYPE=PTR(12), QCLASS=IN(1)
        buf.putShort(12)
        buf.putShort(1)

        return buf.array()
    }

    private fun parsePtrResponse(buf: ByteArray, len: Int, expectedTxId: Int): PtrOutcome {
        if (len < 12) return PtrOutcome.Failed

        val bb = ByteBuffer.wrap(buf, 0, len).order(ByteOrder.BIG_ENDIAN)
        val txId = bb.short.toInt() and 0xFFFF
        if (txId != expectedTxId) return PtrOutcome.Failed

        val flags = bb.short.toInt() and 0xFFFF
        val rcode = flags and 0x000F
        val qdCount = bb.short.toInt() and 0xFFFF
        val anCount = bb.short.toInt() and 0xFFFF

        // Skip questions
        var pos = 12
        for (i in 0 until qdCount) {
            pos = skipName(buf, pos, len)
            if (pos < 0 || pos + 4 > len) return PtrOutcome.Failed
            pos += 4 // QTYPE + QCLASS
        }

        // rcode 3 = NXDOMAIN, treat as authoritative "no record"
        if (rcode == 3) return PtrOutcome.NoRecord
        if (rcode != 0) return PtrOutcome.Failed
        if (anCount == 0) return PtrOutcome.NoRecord

        // Walk answer records, return the first PTR
        for (i in 0 until anCount) {
            pos = skipName(buf, pos, len)
            if (pos < 0 || pos + 10 > len) return PtrOutcome.Failed
            val rrType = ((buf[pos].toInt() and 0xFF) shl 8) or (buf[pos + 1].toInt() and 0xFF)
            val rdLength = ((buf[pos + 8].toInt() and 0xFF) shl 8) or (buf[pos + 9].toInt() and 0xFF)
            pos += 10

            if (pos + rdLength > len) return PtrOutcome.Failed

            if (rrType == 12) { // PTR
                val name = readName(buf, pos, len) ?: return PtrOutcome.Failed
                val cleaned = name.trimEnd('.')
                return if (cleaned.isEmpty()) PtrOutcome.NoRecord else PtrOutcome.Found(cleaned)
            }
            pos += rdLength
        }
        return PtrOutcome.NoRecord
    }

    /** Skip a (possibly compressed) DNS name. Returns the offset past the name, or -1 on error. */
    private fun skipName(buf: ByteArray, start: Int, len: Int): Int {
        var pos = start
        while (pos < len) {
            val b = buf[pos].toInt() and 0xFF
            if (b == 0) return pos + 1
            if ((b and 0xC0) == 0xC0) {
                // pointer — consumes 2 bytes total
                if (pos + 2 > len) return -1
                return pos + 2
            }
            pos += 1 + b
        }
        return -1
    }

    /**
     * Read a DNS name starting at `start`, following compression pointers.
     * Returns the dotted name, or null on malformed input.
     */
    private fun readName(buf: ByteArray, start: Int, len: Int): String? {
        val sb = StringBuilder()
        var pos = start
        var jumped = false
        var hops = 0
        // Cap pointer chasing to defeat malformed loops.
        val maxHops = 32

        while (pos < len) {
            if (hops > maxHops) return null
            val b = buf[pos].toInt() and 0xFF
            if (b == 0) {
                if (!jumped) pos += 1
                break
            }
            if ((b and 0xC0) == 0xC0) {
                if (pos + 1 >= len) return null
                val pointer = ((b and 0x3F) shl 8) or (buf[pos + 1].toInt() and 0xFF)
                if (pointer >= len) return null
                pos = pointer
                jumped = true
                hops++
                continue
            }
            // length-prefixed label
            if (pos + 1 + b > len) return null
            if (sb.isNotEmpty()) sb.append('.')
            sb.append(String(buf, pos + 1, b, Charsets.US_ASCII))
            pos += 1 + b
        }
        return sb.toString()
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // Required for event emitter
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Required for event emitter
    }
}

File: ios/SuperTrace/ReverseDnsModule.swift
import Foundation
import React
import Darwin

@objc(ReverseDnsModule)
class ReverseDnsModule: NSObject, RCTBridgeModule {

    static func moduleName() -> String! {
        return "ReverseDnsModule"
    }

    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    // Public recursive resolvers — first answer wins; we walk this list per-IP.
    private static let dnsServers: [String] = [
        "1.1.1.1",        // Cloudflare
        "8.8.8.8",        // Google
        "9.9.9.9",        // Quad9
        "1.0.0.1",        // Cloudflare secondary
        "8.8.4.4",        // Google secondary
        "208.67.222.222"  // OpenDNS
    ]

    private static let dnsPort: UInt16 = 53
    private static let perServerTimeoutMs: Int = 1500
    private static let totalTimeoutMs: Int = 6000
    private static let maxConcurrent: Int = 12

    // Cap parallel raw-UDP queries so we don't drown the cellular radio.
    private let semaphore = DispatchSemaphore(value: ReverseDnsModule.maxConcurrent)
    private let queue = DispatchQueue(label: "supertrace.rdns", attributes: .concurrent)
    private var txCounter: UInt16 = UInt16.random(in: 0...UInt16.max)
    private let txLock = NSLock()

    private func nextTxId() -> UInt16 {
        txLock.lock()
        defer { txLock.unlock() }
        txCounter = txCounter &+ 1
        return txCounter
    }

    @objc
    func reverseLookup(_ ip: String,
                       resolve: @escaping RCTPromiseResolveBlock,
                       reject: @escaping RCTPromiseRejectBlock) {
        queue.async { [weak self] in
            guard let self = self else { resolve(NSNull()); return }
            self.semaphore.wait()
            defer { self.semaphore.signal() }

            let fqdn = self.performLookup(ip: ip,
                                          deadline: Date().addingTimeInterval(Double(Self.totalTimeoutMs) / 1000.0))
            if let fqdn = fqdn {
                resolve(fqdn)
            } else {
                resolve(NSNull())
            }
        }
    }

    @objc
    func reverseLookupBatch(_ ips: NSArray,
                            resolve: @escaping RCTPromiseResolveBlock,
                            reject: @escaping RCTPromiseRejectBlock) {
        queue.async { [weak self] in
            guard let self = self else { resolve([:]); return }
            let group = DispatchGroup()
            let resultLock = NSLock()
            var result: [String: Any] = [:]

            for case let ip as String in ips {
                group.enter()
                self.queue.async { [weak self] in
                    guard let self = self else { group.leave(); return }
                    self.semaphore.wait()
                    let fqdn = self.performLookup(ip: ip,
                                                  deadline: Date().addingTimeInterval(Double(Self.totalTimeoutMs) / 1000.0))
                    self.semaphore.signal()
                    resultLock.lock()
                    result[ip] = fqdn ?? NSNull()
                    resultLock.unlock()
                    group.leave()
                }
            }

            group.wait()
            resolve(result)
        }
    }

    private enum PtrOutcome {
        case found(String)
        case noRecord
        case failed
    }

    private func performLookup(ip: String, deadline: Date) -> String? {
        guard let arpa = buildArpaName(ip: ip) else { return nil }
        for server in Self.dnsServers {
            if Date() >= deadline { return nil }
            let outcome = queryPtr(server: server, arpaName: arpa)
            switch outcome {
            case .found(let name):
                return name
            case .noRecord:
                return nil
            case .failed:
                continue
            }
        }
        return nil
    }

    private func queryPtr(server: String, arpaName: String) -> PtrOutcome {
        let sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
        if sock < 0 { return .failed }
        defer { close(sock) }

        var tv = timeval()
        tv.tv_sec = Self.perServerTimeoutMs / 1000
        tv.tv_usec = Int32((Self.perServerTimeoutMs % 1000) * 1000)
        setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))
        setsockopt(sock, SOL_SOCKET, SO_SNDTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = Self.dnsPort.bigEndian
        if inet_pton(AF_INET, server, &addr.sin_addr) != 1 {
            return .failed
        }

        let txId = nextTxId()
        let query = buildPtrQuery(txId: txId, arpaName: arpaName)

        let sendResult = query.withUnsafeBytes { (raw: UnsafeRawBufferPointer) -> Int in
            withUnsafePointer(to: &addr) { ptr -> Int in
                ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa -> Int in
                    sendto(sock, raw.baseAddress, raw.count, 0, sa,
                           socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
        }
        if sendResult < 0 { return .failed }

        var buffer = [UInt8](repeating: 0, count: 512)
        var fromAddr = sockaddr_in()
        var fromLen = socklen_t(MemoryLayout<sockaddr_in>.size)
        let received = withUnsafeMutablePointer(to: &fromAddr) { ptr -> Int in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa -> Int in
                recvfrom(sock, &buffer, buffer.count, 0, sa, &fromLen)
            }
        }
        if received <= 0 { return .failed }

        return parsePtrResponse(buffer: buffer, length: received, expectedTxId: txId)
    }

    /// "1.2.3.4" → "4.3.2.1.in-addr.arpa"; IPv6 → 32 reversed nibbles + ".ip6.arpa"
    private func buildArpaName(ip: String) -> String? {
        var v4 = in_addr()
        if inet_pton(AF_INET, ip, &v4) == 1 {
            let raw = withUnsafeBytes(of: &v4.s_addr) { Array($0) }
            // s_addr is in network byte order; bytes are already MSB-first in network terms,
            // but Darwin stores it little-endian on-host. We need the four octets as written
            // ("1.2.3.4") and reverse them. Use inet_ntoa trip to get a canonical form.
            var canonical = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
            inet_ntop(AF_INET, &v4, &canonical, socklen_t(INET_ADDRSTRLEN))
            let canonicalStr = String(cString: canonical)
            let octets = canonicalStr.split(separator: ".")
            if octets.count != 4 { return nil }
            return octets.reversed().joined(separator: ".") + ".in-addr.arpa"
        }

        var v6 = in6_addr()
        if inet_pton(AF_INET6, ip, &v6) == 1 {
            let bytes: [UInt8] = withUnsafeBytes(of: &v6) { raw in
                Array(raw.bindMemory(to: UInt8.self))
            }
            guard bytes.count == 16 else { return nil }
            var sb = ""
            for i in stride(from: 15, through: 0, by: -1) {
                let b = bytes[i]
                sb += String(b & 0x0F, radix: 16) + "."
                sb += String((b >> 4) & 0x0F, radix: 16) + "."
            }
            sb += "ip6.arpa"
            return sb
        }
        return nil
    }

    private func buildPtrQuery(txId: UInt16, arpaName: String) -> [UInt8] {
        let labels = arpaName.split(separator: ".").map { String($0) }
        var out: [UInt8] = []

        // Header: txId, flags=0x0100 (RD), QD=1, AN/NS/AR=0
        out.append(UInt8((txId >> 8) & 0xFF))
        out.append(UInt8(txId & 0xFF))
        out.append(0x01); out.append(0x00)
        out.append(0x00); out.append(0x01)
        out.append(0x00); out.append(0x00)
        out.append(0x00); out.append(0x00)
        out.append(0x00); out.append(0x00)

        for label in labels {
            let bytes = [UInt8](label.utf8)
            out.append(UInt8(bytes.count))
            out.append(contentsOf: bytes)
        }
        out.append(0x00) // root

        // QTYPE=PTR(12), QCLASS=IN(1)
        out.append(0x00); out.append(0x0C)
        out.append(0x00); out.append(0x01)
        return out
    }

    private func parsePtrResponse(buffer: [UInt8], length: Int, expectedTxId: UInt16) -> PtrOutcome {
        if length < 12 { return .failed }
        let txId = (UInt16(buffer[0]) << 8) | UInt16(buffer[1])
        if txId != expectedTxId { return .failed }

        let flags = (UInt16(buffer[2]) << 8) | UInt16(buffer[3])
        let rcode = Int(flags & 0x000F)
        let qdCount = Int((UInt16(buffer[4]) << 8) | UInt16(buffer[5]))
        let anCount = Int((UInt16(buffer[6]) << 8) | UInt16(buffer[7]))

        var pos = 12
        for _ in 0..<qdCount {
            guard let np = skipName(buffer: buffer, start: pos, length: length) else { return .failed }
            pos = np
            if pos + 4 > length { return .failed }
            pos += 4
        }

        if rcode == 3 { return .noRecord }     // NXDOMAIN
        if rcode != 0 { return .failed }
        if anCount == 0 { return .noRecord }

        for _ in 0..<anCount {
            guard let np = skipName(buffer: buffer, start: pos, length: length) else { return .failed }
            pos = np
            if pos + 10 > length { return .failed }
            let rrType = (Int(buffer[pos]) << 8) | Int(buffer[pos + 1])
            let rdLength = (Int(buffer[pos + 8]) << 8) | Int(buffer[pos + 9])
            pos += 10
            if pos + rdLength > length { return .failed }

            if rrType == 12 { // PTR
                guard let name = readName(buffer: buffer, start: pos, length: length) else { return .failed }
                let cleaned = name.trimmingCharacters(in: CharacterSet(charactersIn: "."))
                return cleaned.isEmpty ? .noRecord : .found(cleaned)
            }
            pos += rdLength
        }
        return .noRecord
    }

    private func skipName(buffer: [UInt8], start: Int, length: Int) -> Int? {
        var pos = start
        while pos < length {
            let b = Int(buffer[pos])
            if b == 0 { return pos + 1 }
            if (b & 0xC0) == 0xC0 {
                if pos + 2 > length { return nil }
                return pos + 2
            }
            pos += 1 + b
        }
        return nil
    }

    private func readName(buffer: [UInt8], start: Int, length: Int) -> String? {
        var sb = ""
        var pos = start
        var hops = 0
        let maxHops = 32

        while pos < length {
            if hops > maxHops { return nil }
            let b = Int(buffer[pos])
            if b == 0 { break }
            if (b & 0xC0) == 0xC0 {
                if pos + 1 >= length { return nil }
                let pointer = ((b & 0x3F) << 8) | Int(buffer[pos + 1])
                if pointer >= length { return nil }
                pos = pointer
                hops += 1
                continue
            }
            if pos + 1 + b > length { return nil }
            if !sb.isEmpty { sb.append(".") }
            let labelBytes = Array(buffer[(pos + 1)..<(pos + 1 + b)])
            if let label = String(bytes: labelBytes, encoding: .ascii) {
                sb.append(label)
            } else {
                return nil
            }
            pos += 1 + b
        }
        return sb
    }
}
